event_handling/include/essentials/NotifyTimer.hpp

```diff
@@ -0,0 +1,155 @@
+    ~NotifyTimer();
+    /**
+     * Sets the timer to run.
+     * @return True, if the timer will run after this call. False, otherwise.
+     */
+    bool start();
+    /**
+     * Sets the timer to stop.
+     * @return True, if the timer will anyways run after this call. False, otherwise.
+     */
+    bool stop();
+    /**
+     * Allows to get the running state of the notify timer.
+     * @return True, if the timer is set to run. False, otherwise.
+     */
+    bool isStarted();
+    std::chrono::milliseconds getDelayedStart() const;
+    std::chrono::milliseconds getInterval() const;
+
+private:
+    void runInternal();                        /** < The method executed by the notify timer thread */
+    bool _running;                             /** < Is always true except when the notify timer is shutting down. */
+    bool _started;                             /** < True, if the NotifyTimer is active. False, otherwise. */
+    bool _justStartedAgain;                    /** < True, if the NotifyTimer needs to delay the next call, because it was (re)started. False, otherwise. */
+    std::chrono::milliseconds _msInterval;     /** < The milliseconds between two calls to the callback. */
+    std::chrono::milliseconds _msDelayedStart; /** < The milliseconds between (re)starting the NotifyTimer and the callback call. */
+    std::thread* _runThread;
+    t_notificationcallback<NotificationClass> _callback;
+    NotificationClass* _obj;
+    std::mutex _cv_mtx;
+    std::condition_variable _cv;
+};
+
+template <class NotificationClass>
+NotifyTimer<NotificationClass>::NotifyTimer(std::chrono::milliseconds msInterval, std::chrono::milliseconds msDelayedStart,
+        t_notificationcallback<NotificationClass> callback, NotificationClass* obj)
+        : _msInterval(msInterval)
+        , _msDelayedStart(msDelayedStart)
+        , _running(true)
+        , _started(false)
+        , _justStartedAgain(false)
+{
+    _runThread = new std::thread(&NotifyTimer::runInternal, this);
+    _callback = callback;
+    _obj = obj;
+}
+
+template <class NotificationClass>
+void NotifyTimer<NotificationClass>::runInternal()
+{
+    while (_running) {
+        {
+            std::unique_lock<std::mutex> lck(_cv_mtx);
+            _cv.wait(lck, [&] { return !_running || _started; });
+        }
+
+        if (!_running) // for destroying the timer
+            return;
+
+        if (_justStartedAgain) {
+            if (_msDelayedStart.count() > 0) {
+                std::this_thread::sleep_for(_msDelayedStart);
+            }
+            {
+                std::lock_guard<std::mutex> lockGuard(_cv_mtx);
+                _justStartedAgain = false;
+            }
+        }
+
+        std::chrono::system_clock::time_point start = std::chrono::high_resolution_clock::now();
+        (_obj->*_callback)();
+        auto dura = std::chrono::high_resolution_clock::now() - start;
+        std::this_thread::sleep_for(_msInterval - dura);
+    }
```

Comment:
A shorter version without the need of _justStartedAgain :

            while (_running) {
                {
                    std::unique_lock<std::mutex> lck(_cv_mtx);
                    _cv.wait(lck, [&] { return !_running || _started; });
                }

                if (!_running) // for destroying the timer
                    return;
                
                if (_msDelayedStart.count() > 0) {
                    std::this_thread::sleep_for(_msDelayedStart);
                }

                while(_started) {
                    std::chrono::system_clock::time_point start = std::chrono::high_resolution_clock::now();
                    (_obj->*_callback)();
                    auto dura = std::chrono::high_resolution_clock::now() - start;
                    std::this_thread::sleep_for(_msInterval - dura);
                }                
            }

Reply:
cool